### PR TITLE
`SharedDirectoryCreateRequest` and `SharedDirectoryCreateResponse`

### PR DESCRIPTION
**Pull request by ibeckermayer**
Adds the TDP messages and `SharedDirectoryManager` method necessary to allow users to copy files and directories _into_ the shared directory.

You should now be able to:
- Navigate the directory tree
- Read files
- Write files (< 16kb temporarily)
- Copy files and directories _from_ the shared directory
- Copy files and directories _into_ the shared directory **(new)**
- Move files and directories _into_ the shared directory **(new)**

Moving files and directories _from_ the shared directory still won't work, since it requires unimplemented `SharedDirectoryDelete` messages.

Corresponds with: https://github.com/gravitational/teleport/pull/15289

Requires backport to v9/v10

#### `webapps`
1. Change the value below to `true` https://github.com/gravitational/webapps/blob/becfad38776a68e0d8be634476f24cead7890ffc/packages/teleport/src/config.ts#L32
2. run the development server like `yarn start-teleport --target=https://ec2-35-171-27-185.compute-1.amazonaws.com/`
3. See https://gravitational.slack.com/archives/C02DQ1C2BMW/p1657807702085829 for a username/password combo

## What to expect

Go to the top right `...` menu and click `Share Directory`

image

(X-out of the "uncaught exception" screen if you get it). The directory you select should show up in the File Explorer as a shared drive

Screen Shot 2022-07-19 at 14 13 45

The drive should be navigable as well should any folders within it

image

### CODE REVIEW

File: packages/teleport/src/lib/tdp/client.ts [hunk 1]
```diff
@@ -136,6 +137,14 @@ export default class Client extends EventEmitterWebAuthnSender {
         case MessageType.SHARED_DIRECTORY_INFO_REQUEST:
           this.handleSharedDirectoryInfoRequest(buffer);
           break;
+        case MessageType.SHARED_DIRECTORY_CREATE_REQUEST:
+          // A typical sequence is that we receive a SharedDirectoryCreateRequest
+          // immediately followed by a SharedDirectoryWriteRequest. It's important
+          // that we await here so that this client doesn't field the SharedDirectoryWriteRequest
+          // until the create has successfully completed, or else we might get an error
+          // trying to write to a file that hasn't been created yet.
+          await this.handleSharedDirectoryCreateRequest(buffer);
```
Comment (hatched) [inline packages/teleport/src/lib/tdp/client.ts:146]: nice comment, thx

File: packages/teleport/src/lib/tdp/client.ts [hunk 2]
```diff
@@ -102,7 +103,7 @@ export default class Client extends EventEmitterWebAuthnSender {
     };
   }
 
-  processMessage(buffer: ArrayBuffer) {
+  async processMessage(buffer: ArrayBuffer) {
```
Comment (hatched) [inline packages/teleport/src/lib/tdp/client.ts:106]: now that this is an `async` function you should probably type the return value here.
```suggestion
  async processMessage(buffer: ArrayBuffer): Promise<void> {
```

File: packages/teleport/src/lib/tdp/playerClient.ts [hunk 3]
```diff
@@ -42,7 +42,7 @@ export class PlayerClient extends Client {
   }
 
   // Overrides Client implementation.
-  processMessage(buffer: ArrayBuffer) {
+  async processMessage(buffer: ArrayBuffer) {
```
Comment (hatched) [inline packages/teleport/src/lib/tdp/playerClient.ts:45]: See comment above about the return type.

File: packages/teleport/src/lib/tdp/sharedDirectoryManager.ts [hunk 4]
```diff
@@ -11,6 +11,7 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
+import { FileType } from './codec';
```
Comment (hatched) [inline packages/teleport/src/lib/tdp/sharedDirectoryManager.ts:14]: Plz be explicit when importing types.

```suggestion
import type { FileType } from './codec';
```

**Review by hatched — APPROVED** [review]

File: packages/teleport/src/lib/tdp/client.ts (hunk 2 above)
Comment (ibeckermayer) [reply packages/teleport/src/lib/tdp/client.ts:106]: Why is that particularly important for `async` functions?

File: packages/teleport/src/lib/tdp/sharedDirectoryManager.ts (hunk 4 above)
Comment (ibeckermayer) [reply packages/teleport/src/lib/tdp/sharedDirectoryManager.ts:14]: It's used as an object here

https://github.com/gravitational/webapps/blob/189c1d4ed84ed7a282c7028239ac3ce199f4e668/packages/teleport/src/lib/tdp/sharedDirectoryManager.ts#L166

Comment (hatched) [reply packages/teleport/src/lib/tdp/sharedDirectoryManager.ts:14]: Ahh I see now, thanks sorry I missed that.

File: packages/teleport/src/lib/tdp/client.ts (hunk 2 above)
Comment (hatched) [reply packages/teleport/src/lib/tdp/client.ts]: async functions have an implicit return type of a `Promise` but the value from within that promise isn't automagically picked up by the compiler so by specifying the type in the generic there it just helps it be more explicit in the editor.

Comment (JanKaczmarkiewicz) [reply packages/teleport/src/lib/tdp/client.ts]: > value from within that promise isn't automagically picked up by the compiler

it is actually:
Screenshot 2022-08-09 at 18 13 52

Comment (hatched) [reply packages/teleport/src/lib/tdp/client.ts]: Oh nice, it wasn't working for me in this PR so that's why I suggested the change.

**Pull request merged**